### PR TITLE
Initial support for easyname as registrar

### PR DESCRIPTION
**Pull request by tresni**
Started working on a Cloudflare registrar, but official API docs indicate the necessary information isn't available via their API.  So I did a registrar provider for [easyname](https://easyname.com) for the time being.

No official golang API library for them, so wrote this all from scratch based on the [easyname API documentation](https://api-docs.easyname.com/) and the [easyname php-sdk](https://github.com/easyname/php-sdk/).  Tested against my personal account

### CODE REVIEW

**Review by tlimoncelli — COMMENTED** [review]
Looks good so far. I'll have more time to review this later this weekend.

My one global suggestion is to move the protocol stuff to a separate file.

**Comment by tresni** [comment]
@tlimoncelli I thought about it and then didn't.  I've gone ahead and separated the provider logic from the api logic.

File: providers/easyname/provider.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+func init() {
+	providers.RegisterRegistrarType("EASYNAME", newEasyname)
+}
+
+func newEasyname(m map[string]string) (providers.Registrar, error) {
+	api := &easynameProvider{}
+
+	if m["email"] == "" || m["userid"] == "" || m["apikey"] == "" || m["authsalt"] == "" || m["signsalt"] == "" {
+		return nil, fmt.Errorf("missing easyname email, userid, apikey, and/or salt")
```
Comment (tlimoncelli) [inline providers/easyname/provider.go:27]: error message should list authsalt, not salt.

File: providers/easyname/provider.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+			nss = append(nss, ns)
+		}
+	}
+	foundNameservers := strings.Join(nss, ",")
+
+	expected := []string{}
+	for _, ns := range dc.Nameservers {
+		expected = append(expected, ns.Name)
+	}
+	sort.Strings(expected)
```
Comment (tlimoncelli) [inline providers/easyname/provider.go:56]: Does found need to be sorted too? (add a comment if not)

File: providers/easyname/api.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+	er := easynameResponseData{}
+	if err = c.request("POST", fmt.Sprintf("https://api.easyname.com/domain/%d/nameserverchange", domain), bytes.NewBuffer(body), &er); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *easynameProvider) getDomain(domain string) (easynameDomain, error) {
```
Comment (tlimoncelli) [inline providers/easyname/api.go:116]: getDomain should be getDomains or getDomainList

File: providers/easyname/provider.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+func newEasyname(m map[string]string) (providers.Registrar, error) {
+	api := &easynameProvider{}
+
+	if m["email"] == "" || m["userid"] == "" || m["apikey"] == "" || m["authsalt"] == "" || m["signsalt"] == "" {
+		return nil, fmt.Errorf("missing easyname email, userid, apikey, and/or salt")
+	}
+
+	api.apikey, api.signSalt = m["apikey"], m["signsalt"]
+	composed := fmt.Sprintf(m["authsalt"], m["userid"], m["email"])
```
Comment (tlimoncelli) [inline providers/easyname/provider.go:31]: Missing format string.

File: providers/easyname/provider.go (hunk 1 above)
Comment (tresni) [reply providers/easyname/provider.go:27]: Good catch, that was from before I added `signsalt` so `authsalt` was just `salt`.  Needs to include both of them.

File: providers/easyname/api.go (hunk 3 above)
Comment (tresni) [reply providers/easyname/api.go:116]: This returns a single domain's nameservers so getDomains would not be correct (and that's the function fetchDomainList below.)  Might be better to name it getNameservers (to match updateNameservers above) or getDomainNameservers?

File: providers/easyname/provider.go (hunk 4 above)
Comment (tresni) [reply providers/easyname/provider.go:31]: No, `authsalt` is the format string ;-)  It's a really weird system.

File: providers/easyname/api.go (hunk 3 above)
Comment (tresni) [reply providers/easyname/api.go:116]: Error message needs updated though

**Comment by tresni** [comment]
@tlimoncelli I've addressed all your comments and merged in latest master.  Please let me know if you want to see any other changes.

**Comment by tlimoncelli** [comment]
Two easy things to fix:  Update README.md and OWNERS.  See https://stackexchange.github.io/dnscontrol/writing-providers for details.

**Comment by tresni** [comment]
@tlimoncelli I didn't like the idea of calling it a DNS Provider, so I moved it to a new section called "Domain Registrars" and populated that with all providers that act as domain registrar providers.

**Comment by tlimoncelli** [comment]
> @tlimoncelli I didn't like the idea of calling it a DNS Provider, so I moved it to a new section called "Domain Registrars" and populated that with all providers that act as domain registrar providers.

Excellent idea.  We've actually had confusion about that in the past.

I'll merge it shortly.

**Pull request merged**